scipy-stubs/signal/_peak_finding.pyi

```diff
@@ -1,37 +1,86 @@
-from scipy._typing import Untyped
+from collections.abc import Callable, Sequence
+from typing import Concatenate, Literal, TypeAlias, TypedDict
+
+import numpy as np
+import numpy.typing as npt
+import optype as op
+from numpy._typing import _ArrayLikeFloat_co, _ArrayLikeInt_co
+from scipy._typing import AnyInt, AnyReal
 
 __all__ = ["argrelextrema", "argrelmax", "argrelmin", "find_peaks", "find_peaks_cwt", "peak_prominences", "peak_widths"]
 
-def argrelmin(data: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def argrelmax(data: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def argrelextrema(data: Untyped, comparator: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def peak_prominences(x: Untyped, peaks: Untyped, wlen: Untyped | None = None) -> Untyped: ...
+_Array_n: TypeAlias = npt.NDArray[np.intp]
+_Array_n_1d: TypeAlias = np.ndarray[tuple[int], np.dtype[np.intp]]
+_Array_f8: TypeAlias = npt.NDArray[np.float64]
+_Mode: TypeAlias = Literal["clip", "wrap"]
+
+_ProminencesResult: TypeAlias = tuple[_Array_f8, _Array_n, _Array_n]
+_WidthsResult: TypeAlias = tuple[_Array_f8, _Array_f8, _Array_f8, _Array_f8]
+
+class _FindPeaksResultsDict(TypedDict, total=False):
+    peak_heights: _Array_f8
+    left_thresholds: _Array_f8
+    right_thresholds: _Array_f8
+    prominences: _Array_f8
+    left_bases: _Array_n
+    right_bases: _Array_n
+    width_heights: _Array_f8
+    left_ips: _Array_f8
+    right_ips: _Array_f8
+    plateau_sizes: _Array_n
+    left_edges: _Array_n
+    right_edges: _Array_n
+
+def argrelmin(
+    data: npt.NDArray[np.generic],
+    axis: op.CanIndex = 0,
+    order: int = 1,
+    mode: _Mode = "clip",
+) -> tuple[_Array_n, ...]: ...
+def argrelmax(
+    data: npt.NDArray[np.generic],
+    axis: op.CanIndex = 0,
+    order: int = 1,
+    mode: _Mode = "clip",
+) -> tuple[_Array_n, ...]: ...
+def argrelextrema(
+    data: npt.NDArray[np.generic],
+    comparator: Callable[[npt.NDArray[np.generic], npt.NDArray[np.generic]], npt.NDArray[np.bool_]],
```

Comment:
totally optional; but you could use a `TypeVar(..., bound=np.generic)` instead of the current `np.generic`s

Reply:
Wouldn't that require the comparator function to have two arguments of the same type? I think that in the ideal case the function signature makes sense but what if someone were to pass a function that takes in a float and an int? They would still be able to do the comparison despite the types being different.

Assuming that using `TypeVar` enforces that the two argument types to be the same that is.

Reply:
> Wouldn't that require the comparator function to have two arguments of the same type? 

Yes exactly. 

---

> I think that in the ideal case the function signature makes sense but what if someone were to pass a function that takes in a float and an int? 

If someone would pass a function that accepts arrays whose scalar types are incompatible, then it would indeed not be allowed. 
And that's a good thing, because it could prevent some potentially type-unsafe situations.

> Assuming that using `TypeVar` enforces that the two argument types to be the same that is.

Well, it kinda depends on what you mean with "the same", because in this case, we also need to take the relationship of `NDArray` and its scalar type parameter into account, which is a covariant one.
In this case that means that you're allowed to pass something like `(NDArray[np.int32], NDArray[np.integer[Any]]) -> NDArray[np.bool_]` as function if you also pass `data: NDarray[np.int32]`.

For a demonstration of this, see https://basedpyright.com/?pythonVersion=3.13&typeCheckingMode=all&reportUnusedExpression=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUkBHAVySxiQChRIpUBPCuiwBzMpWq0AwgUTFkAGlgEcqTpwDEULGFRI%2BACwK0CMOCAAmwsajB9BKAygDWSEDiUADYak9kcUCo46CJY8vq2UJ7A8GDGnuoCFPoAgiAgBPwA2gAqALpQALz0TKzsSLkFUFpwAG4EIOgEWLQAFADu1M44AJSa9slQaRnZ%2BUUlLGwclVDVZFj1jc1tFmBIOFgA5LSdIM59nBZIwFDAla0WxgQAXEPpmZVKcJQNxtS3MnJEyFlZww-5JT-UZ5PJA%2B7ZIhgBCgnpQAC0AD4oFCENdOLNZiAkKhmCAsIFEK1nhRXrYQBcrllrvCAIxgqCXVAELK0655HoHI4nHwAfWQvJicVQrRudxGWR8DKIt2BWSFxgZAHo4UjxQ9UfA8ujMVBsbj8VAsgR0FAADzFIim4DUQLoJRW%2BZQABe6AooodPTy6kOV3GMBUIrlUqU8HQqlaGVESFaAFZOX1w7ycJIkBZxsBKcylHyBQrUAcgA

Reply:
Ah got it. Just misunderstood the semantics. I can add this then.